### ISSUE TITLE
Help - Trying to update from ES5 to ES6+ and losing templateContext during rendering

### ISSUE THREAD

**Issue by attitudemattrs** [marionettejs/backbone.marionette#3708]
I apologize up front if this is not the correct forum to post this particular question, but I am hoping someone here can help me with my current predicament.

We have a backbone.marionette application written using TypeScript with a target of ES5. Up to this point, the framework has worked well for us.  For the last several days I have been attempting to upgrade our JavaScript target from ES5 to ES6+ (i.e., es2020) and have run into the unfortunate circumstance that apparently backbone.marionette does not work well within an ES6+ context.

I have read through several discussions on this topic, including the two below:

https://benmccormick.org/2015/04/06/221944.html
https://benmccormick.org/2015/07/06/062222.html

Unfortunately, I have not been able to come up with a solution to our specific problem, which is that the render does not have access
to the `templateContext` following the call to `mixinTemplateContext`.  If I understand the situation correctly, this is due to the restrictions on modifying ES6+ classes after construction.  Note, I am not well versed on this topic.

An example of our situation is as follows.  We have the following View (PageHeaderItemView) defined. The options object includes the `model` for the view (set prior to construction), the `template` method, `tagName` attribute, and `templateContext` attribute. `ViewHelpers.viewHelper` is a template helper object with numerous helper functions in it.

```
   export class PageHeaderItemView extends Marionette.View<PageHeaderModel>
   {
	:
      constructor(options?:any)
      {
         if (!options)
            options = {};

         options.template        = Marionette.TemplateCache.get("#pageHeaderTemplate");
         options.tagName         = "div";
         options.templateContext = ViewHelpers.viewHelper;
         super(options);
		:
      }
   }

```
When the `PageHeaderItemView` view is rendered, within template-render.js, the following method is called:

```
  mixinTemplateContext(serializedData) {
    const templateContext = _.result(this, 'templateContext');
    if (!templateContext) { return serializedData; }
    if (!serializedData) { return templateContext; }
    return _.extend({}, serializedData, templateContext);
  },

```
Within this method we have a valid `templateContext` object (our helper methods object) and a valid `serializedData` object (our model object). With ES5 this works as expected. However with ES6 and above, this does not work. When the method completes, the resulting object represents the model, but does not include the `templateContext`'s methods.

As noted, I have read several discussions regarding these types of issues within backbone/mariontette with ES6+, but none of them have provided me with a solution for the loss of the `templateContext `(helpers) when the view is rendered.

I'm hoping someone here can provide me with a possible solution or some alterative options. Otherwise, I fear that the only alternative for us is to stay with ES5 and start looking at moving to a different framework that has support for ES6+ classes in the future.

Any help would be greatly appreciated.

**Comment by Upperstage** [comment]
/**
 * @extends Marionette.View
 * @desc The view used render subject results in the read-only list of subjects
 * @since 4 Jun 2023
 */
export default class DurationResultsItemView extends Marionette.View<SubjectResultsModel> {
    private _customizationOptions: any;

    /**
     * @constructor
     * @override
     * @desc Typescript complains when certain properties are added to the class as methods (even though
     * Backbone supports it), and Backbone doesn't see them properties when they're added via Typescript.
     * Also, doesn't make sense for caller to provide this stuff.
     */
    constructor(options: any) {
        super(
            Object.assign({}, options, {
                className: () => 'data-row',

                template: MustacheTemplate,
                
                // Returning the template context like this establishes this
                templateContext: function (this: DurationResultsItemView) {
                    return Object.assign({}, this.model.attributes, DurationTemplateContext, {
                        customizationOptions: this.customizationOptions,
                    });
                },
            })
        );

        this._customizationOptions = options;
    }

    /**
     * Simple getter for template functions
     */
    public get customizationOptions(): any {
        return this._customizationOptions;
    }
}

**maybe this helps?**

**Comment by attitudemattrs** [comment]
Thank you very much for your reply.  Great suggestion.  Unfortunately, I don't think this helps. The `Object.assign` only copies enumerable own properties, which doesn't include the object's methods, just the attributes.  I did try it just to be sure, but the methods were not included.  Thanks again for the suggestion!

**Comment by attitudemattrs** [comment]
You may be right.  I have tried setting `templateContext = { foo() { return "bar" } }` as you have done or using a locally defined class `class Helper { foo() {  return "bar" } }` and setting `templateContext = new Helper()` and both work in the fiddle and in my application.  However, within my TypeScript application the `ViewHelpers.viewHelper` class instance is not working (we have a module ViewHelpers that defines a class ViewHelper and this is an instance of that class) .  This is working with ES5 but not with ES6+.  I just need to figure out why.  If I can figure out how to create a fiddle of this, I will.

Sorry for any wasted time.  I appreciate the responses thus far.  I'll keep looking.

Thanks.

**Comment by attitudemattrs** [comment]
I just thought I would circle back on this to let you know I found a solution to my issue.  You were correct, it was not an issue specifically with Marionette.  The issue was with the difference between using ES6 JavaScript object instances and plain objects, in conjunction with Backbone/Marionette.  After testing with your fiddle example I realized that if I used plain objects (as demonstrated in the fiddle example and my local testing) that things seemed to work. So, I went in search of a solution to convert our ES6+ JavaScript object instances to a plain objects.  It appears I have found a solution that works for my situation.  The relevant post was here:

https://stackoverflow.com/questions/34699529/convert-javascript-class-instance-to-plain-object-preserving-methods

In particular, the solution that worked for me was a method posted by **eAbi** on Jan 10, 2016.  The solution is midway down the page and refers to a method called `toJSON(proto)`.  Method naming aside, the function converts an ES6 JavaScript object instance into a plain object, taking into consideration inheritance.  

Using this function, I set my `templateContext` as follows:

`options.templateContext = ViewHelpers.ViewHelperConverter.toJSON(ViewHelpers.viewHelper);
`

With this done, I am able to build my TypeScript project  using backbone.marionette v4.1.2 and backbone 1.4.1 with a target of ES2020 and run it, with everything thus far appearing to work as it did before with ES5.  

Though I have much more testing to do, this appears to resolve my issue.  Closing the issue.

I appreciate your time and efforts.
